utils.groovy

```diff
@@ -282,4 +282,110 @@ def get_artifacts_to_build(pipecfg, stream, basearch) {
     }
     return artifacts.unique()
 }
+// Upload artifacts to clouds
+def upload_to_clouds(pipecfg, basearch, buildID, stream) {
+    // Get artifacts to upload
+    def meta_json = "builds/${buildID}/${basearch}/meta.json"
+    def meta = readJSON file: meta_json
+    def artifacts = meta['images'].keySet() as List
+    def clouds = pipecfg.clouds.keySet() as List
+    // Check if we have cloud config to upload these artifacts
+    def uploads = artifacts.intersect(clouds)
+    parallel uploads.inject([:]) { branch, cloud -> branch["Upload ${cloud}"] = {
```

Comment:
Just to note here some of the OOB discussions we had about this. I think rather than trying to make a single closure be generic, you can have multiple separate closures for each cloud. The idea I had was doing something like this:

```groovy
def uploaders = [:]
uploaders["aws"] = {
  ...
}
uploaders["gcp"] = {
  ...
}
...
// use actual artifacts list to filter out uploaders we don't need
// then:
parallel uploaders
```

Also, if possible I think it'd be good to move this to a dedicated module, e.g. `uploads.groovy`?

In the future, I think it'd be pretty cool to be able to just do `cosa buildextend-${cloud} --config config.yaml`. Then we could do something similar to what you originally had in mind here.

Reply:
>> I think rather than trying to make a single closure be generic, you can have multiple separate closures for each cloud
In this case we would lose the `tryWithCredentials` generic utilization and would need to call it for all + the generic utilization for `aws` and `azure` that are almost identical.  Is that what we want?